ENIGMAsystem/SHELL/Platforms/General/UNIXfilemanip.cpp

```diff
@@ -20,48 +20,626 @@
+        S_ISREG(sb.st_mode));
+}
+
+int file_delete(std::string fname)
+{
+    std::string str(fname);
+
+    struct stat sb;
+    if (stat((char *)str.c_str(),&sb) == 0 &&
+        S_ISREG(sb.st_mode))
+    {
+        return (remove((char *)str.c_str()) == 0);
+    }
+
+    return 0;
+}
+
+int file_rename(std::string oldname,std::string newname)
+{
+    std::string str1(oldname);
+    std::string str2(newname);
+
+    str1.erase(unique(str1.begin(),str1.end(),dupslash()),str1.end());
+    str2.erase(unique(str2.begin(),str2.end(),dupslash()),str2.end());
+
+    size_t slash1;
+    size_t slash2;
+    std::string dir1(str1);
+    std::string dir2(str2);
+
+    if (dir1.find('/') != std::string::npos)
+        slash1 = dir1.find_last_of("/");
+    else
+    {
+        dir1 = std::string("");
+        slash1 = 0;
+    }
+
+    if (dir2.find('/') != std::string::npos)
+        slash2 = dir2.find_last_of("/");
+    else
+    {
+        dir2 = std::string("");
+        slash2 = 0;
+    }
+
+    int proceed = (dir1.substr(0,slash1) != dir2.substr(0,slash2));
+
+    std::string prestr1 = str1;
+    std::string prestr2 = str2;
+
+    std::string pardir1(dirname((char *)str1.c_str()));
+    if (!pardir1.empty())
+    {
+        while (*pardir1.rbegin() == '/')
+        {
+            pardir1.erase(pardir1.size()-1);
+        }
+    }
+    struct stat pexist1;
+    if (stat((char *)pardir1.c_str(),&pexist1) == 0 &&
+        S_ISDIR(pexist1.st_mode))
+    {
+        char actpath1[PATH_MAX+1];
+        char *ptr1 = realpath((char *)pardir1.c_str(),actpath1);
+        if (ptr1 != NULL)
+        {
+            std::string pardir2(dirname((char *)str2.c_str()));
+            if (!pardir2.empty())
+            {
+                while (*pardir2.rbegin() == '/')
+                {
+                    pardir2.erase(pardir2.size()-1);
+                }
+            }
+            struct stat pexist2;
+            if (stat((char *)pardir2.c_str(),&pexist2) == 0 &&
+                S_ISDIR(pexist2.st_mode))
+            {
+                char actpath2[PATH_MAX+1];
+                char *ptr2 = realpath((char *)pardir2.c_str(),actpath2);
+                if (ptr2 != NULL)
+                {
+                    proceed = (std::string(actpath1) != std::string(actpath2));
+                }
+            }
+        }
+    }
+
+    str1 = prestr1;
+    str2 = prestr2;
+
+    if (proceed || basename((char *)str1.c_str()) != basename((char *)str2.c_str()))
+    {
+        struct stat sb1;
+        struct stat sb2;
+        if (stat((char *)str1.c_str(),&sb1) == 0 &&
+            S_ISREG(sb1.st_mode) &&
+            stat((char *)str2.c_str(),&sb2) != 0)
+        {
+            return (rename((char *)str1.c_str(),(char *)str2.c_str()) == 0);
+        }
+    }
+
+    return 0;
 }
 
-int file_delete(string fname)
+int file_copy(std::string fname,std::string newname)
 {
-  return remove(fname.c_str());
+    std::string str1(fname);
```

Comment:
Here and throughout: please extract path logic into helper functions, or use the existing ones (`filename_name`, `filename_path`, etc). This will greatly improve readability and maintainability. Copying a file is actually a hard problem if you want to avoid overwriting existing files and would like to handle links properly. I'm not sure why there are no POSIX shorthands for it.